File: Chapter_01_Introduction/Getting_Started_The_Easy_Problems/11679-Sub_Prime/main.cpp
/*
 * 11679 - Sub-prime
 * debtor is -
 * creditor is +
 */
#include <iostream>
#include <sstream>
#include <vector>

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    std::ostringstream output;
    int numberOfBanks {}, numberOfDebentures {};
    while (std::cin >> numberOfBanks >> numberOfDebentures && numberOfBanks != 0
            && numberOfDebentures != 0) {
        std::vector<int> banks (numberOfBanks + 1);
        for (int i {1}; i < numberOfBanks + 1; ++i) {
            std::cin >> banks[i];
        }
        for (int i {}; i < numberOfDebentures; ++i) {
            int debtor{}, creditor, value;
            std::cin >> debtor >> creditor >> value;
            banks[debtor] -= value;
            banks[creditor] += value;
        }
        bool negativeBalance {};
        for (int i {1}; i < numberOfBanks + 1; ++i) {
            if (banks[i] < 0) {
                negativeBalance = true;
                break;
            }
        }
        output << (negativeBalance ? "N\n" :"S\n");
    }
    std::cout << output.str();
}

File: Chapter_01_Introduction/Getting_Started_The_Easy_Problems/11679-Sub_Prime/Main.java
/*
 * Problem 11679 - Sub-prime
 * Java version 6/12/2016
 */

import java.io.PrintStream;
import java.util.Scanner;

public class Main {

    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);
        StringBuffer results = new StringBuffer();

        while (true) {
            int numberOfBanks = input.nextInt();
            int numberOfDebentures = input.nextInt();

            if (numberOfBanks == 0 && numberOfDebentures == 0) {
                break;
            }

            int[] a = new int[numberOfBanks];
            for (int i = 0; i < a.length; ++i) {
                a[i] = input.nextInt();
            }

            for (int i = 0; i < numberOfDebentures; ++i) {
                int debtor = input.nextInt() - 1;
                int creditor = input.nextInt() - 1;
                int value = input.nextInt();

                a[debtor] -= value;
                a[creditor] += value;
            }

            boolean solvant = true;
            for (int i = 0; i < a.length; ++i) {
                if (a[i] < 0) {
                    solvant = false;
                    break;
                }
            }
            if (solvant) {
                results.append("S\n");
            } else {
                results.append("N\n");
            }

        }
        System.out.print(results.toString());
    }
}

